### PR TITLE
Work around broken enum functions run_time_cache on PHP 8.2.0 and 8.2.1

### LINKED ISSUE #1859: [BUG] SIGSEGV when using backed enums in PHP 8.2
### Bug description
When using the `from` method on a backed enum in PHP 8.2 while the DataDog tracer is enabled, I get a SIGSEGV starting from the second call to fpm.

A full reproducible case can be found here: https://github.com/JeppeKnockaert/datadog-sigsegv-reproduction

Run `docker compose up`, navigate to http://localhost and refresh the page.

### PHP version
PHP 8.2.1 (cli) (built: Jan  6 2023 15:30:34) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.2.1, Copyright (c) Zend Technologies
    with Zend OPcache v8.2.1, Copyright (c), by Zend Technologies
    with ddtrace v0.83.0, Copyright Datadog, by Datadog
    with ddappsec v0.4.5, Copyright Datadog, by Datadog

### Tracer version
0.83.0

### Installed extensions
[PHP Modules]
Core
date
ddappsec
ddtrace
filter
hash
json
libxml
pcre
random
readline
Reflection
SPL
standard
Zend OPcache
zlib

[Zend Modules]
Zend OPcache
ddappsec
ddtrace

### OS info
NAME="Alpine Linux"
VERSION_ID=3.17.1
PRETTY_NAME="Alpine Linux v3.17"

#### Output of phpinfo() (ddtrace >= 0.47.0)

Screenshot 2023-01-16 at 16 04 27
Screenshot 2023-01-16 at 16 04 37
Screenshot 2023-01-16 at 16 04 50
Screenshot 2023-01-16 at 16 05 01
Screenshot 2023-01-16 at 16 05 12
Screenshot 2023-01-16 at 16 05 40

### Core dump

```
Reading symbols from /usr/sbin/php-fpm82...
Reading symbols from /usr/lib/debug//usr/sbin/php-fpm82.debug...

warning: Can't open file /dev/zero (deleted) during file-backed mapping note processing
[New LWP 26]
[New LWP 74]
Core was generated by `php-fpm: pool www                                                       '.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000561a8be4aa53 in _zend_observe_fcall_begin (execute_data=0x7f1e6fa150a0) at Zend/zend_observer.c:245
245	Zend/zend_observer.c: No such file or directory.
[Current thread is 1 (LWP 26)]
(gdb) bt
#0  0x0000561a8be4aa53 in _zend_observe_fcall_begin (execute_data=0x7f1e6fa150a0) at Zend/zend_observer.c:245
#1  0x0000561a8bde675b in ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (execute_data=0x7f1e6fa15020) at Zend/zend_vm_execute.h:2058
#2  0x0000561a8bdc9808 in execute_ex (ex=<optimized out>) at Zend/zend_vm_execute.h:55807
#3  0x0000561a8bdc99ee in zend_execute (op_array=op_array@entry=0x7f1e6fa02000, return_value=return_value@entry=0x0) at Zend/zend_vm_execute.h:60380
#4  0x0000561a8bda5d6b in zend_execute_scripts (type=type@entry=8, retval=retval@entry=0x0, file_count=file_count@entry=3) at Zend/zend.c:1780
#5  0x0000561a8bd3c167 in php_execute_script (primary_file=primary_file@entry=0x7fff237b8058) at main/main.c:2540
#6  0x0000561a8be98c27 in main (argc=5, argv=0x7fff237b82c8) at sapi/fpm/fpm/fpm_main.c:1891
```

### Valgrind output

```
datadog-sigsegv-reproduction-fpm-1    | ==7== Conditional jump or move depends on uninitialised value(s)
datadog-sigsegv-reproduction-fpm-1    | ==7==    at 0x552931: _zend_observe_fcall_begin (zend_observer.c:228)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x4EE75A: ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (zend_vm_execute.h:2058)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x4D1807: execute_ex (zend_vm_execute.h:55807)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x4D19ED: zend_execute (zend_vm_execute.h:60380)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x4ADD6A: zend_execute_scripts (zend.c:1780)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x444166: php_execute_script (main.c:2540)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x5A0C26: main (fpm_main.c:1891)
datadog-sigsegv-reproduction-fpm-1    | ==7==  Uninitialised value was created by a heap allocation
datadog-sigsegv-reproduction-fpm-1    | ==7==    at 0x48A5640: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x47DD18: __zend_malloc (zend_alloc.c:3114)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x486167: zend_arena_alloc (zend_arena.h:67)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x486167: zend_compile_class_decl (zend_compile.c:7817)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x483A23: zend_compile_top_stmt (zend_compile.c:10079)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x4839D9: zend_compile_top_stmt (zend_compile.c:10068)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x4887EC: zend_compile_namespace (zend_compile.c:8308)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x4887EC: zend_compile_stmt (zend_compile.c:10184)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x483A34: zend_compile_top_stmt (zend_compile.c:10082)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x4839D9: zend_compile_top_stmt (zend_compile.c:10068)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x468DB8: zend_compile (zend_language_scanner.c:620)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x468C69: compile_file (zend_language_scanner.c:655)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x50A85CC: _dd_compile_file (engine_hooks.c:90)
datadog-sigsegv-reproduction-fpm-1    | ==7==    by 0x4F7F3D6: opcache_compile_file (ZendAccelerator.c:1820)
datadog-sigsegv-reproduction-fpm-1    | ==7==
```

### LINKED ISSUE #1860: [BUG] RoadRunner integration create traces with TraceId 0, rejected by APM agent
### Bug description

While trying out the new RoadRunner integration, the generated traces are rejected by the APM agent with the following error message: 

```
023-01-16 14:31:33 UTC | TRACE | DEBUG | (run.go:248 in Debugf) | Dropping invalid trace: TraceID is zero (reason:trace_id_zero): service:"roadrunner" name:"web.request" resource:"GET /api/monitoring/status/liveness" spanID:4833514502463363744 start:1673879492413635000 duration:3026000 meta:<key:"_dd.p.dm" value:"-1" > meta:<key:"http.method" value:"GET" > meta:<key:"http.status_code" value:"200" > meta:<key:"http.url" value:"http://127.0.0.1:8000/api/monitoring/status/liveness" > meta:<key:"http.useragent" value:"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36" > meta:<key:"span.kind" value:"server" > metrics:<key:"_dd.rule_psr" value:1 > metrics:<key:"_sampling_priority_v1" value:1 > metrics:<key:"php.compilation.total_time_ms" value:149.706 > metrics:<key:"process_id" value:13272 > type:"web"
```

I've narrowed it down to here: https://github.com/DataDog/dd-trace-php/blob/master/ext/ddtrace.c#L1867 where `DDTRACE_G(distributed_trace_id)` is `NULL`

Reproduction case with

```
docker-compose run --rm 8.0-buster bash
```

and tweaking to start an agent in the compose environment.

```
cat << EOF > .rr.yaml
server:
  command: "php -dlog_errors='on' -ddatadog.trace.client_ip_header_disabled='true' -derror_log='/tmp/dd-trace-php/tests/Frameworks/Roadrunner/Version_2/dd_php_error.log' -dddtrace.request_init_hook='/tmp/dd-trace-php/bridge/dd_wrap_autoloader.php' -dxdebug.remote_enable='1' -dxdebug.remote_host='host.docker.internal' -dxdebug.remote_autostart='1' /tmp/dd-trace-php/tests/Frameworks/Roadrunner/Version_2/worker.php"
  env: {"DD_TRACE_AGENT_PORT":8126,"DD_AGENT_HOST":"agent","DD_TRACE_AGENT_FLUSH_AFTER_N_REQUESTS":1,"DD_TRACE_AGENT_FLUSH_INTERVAL":333,"DD_TEST_INTEGRATION":"true","DD_AUTOLOAD_NO_COMPILE":"true","DD_TRACE_DEBUG":"false","DD_TRACE_HEADER_TAGS":"x-header","DD_TRACE_CLI_ENABLED":1}
  exec_timeout: 5s
http:
  address: 0.0.0.0:9999
  pool:
    num_workers: 1
    destroy_timeout: 1s
logs:
  output: /tmp/dd-trace-php/tests/Frameworks/Roadrunner/Version_2/roadrunner_error.log
  err_output: /tmp/dd-trace-php/tests/Frameworks/Roadrunner/Version_2/roadrunner_error.log
EOF
```

Then if you `curl http://127.0.0.1:9999`, we see in the agent logs:

```
2023-01-16 15:59:03 UTC | TRACE | DEBUG | (run.go:248 in Debugf) | Dropping invalid trace: TraceID is zero (reason:trace_id_zero): service:"roadrunner" name:"web.request" resource:"GET /" spanID:2830416086992939982 start:1673884742821847352 duration:211734 meta:<key:"_dd.p.dm" value:"-1" > meta:<key:"http.method" value:"GET" > meta:<key:"http.status_code" value:"200" > meta:<key:"http.url" value:"http://127.0.0.1:9999/" > meta:<key:"http.useragent" value:"curl/7.64.0" > meta:<key:"span.kind" value:"server" > metrics:<key:"_dd.rule_psr" value:1 > metrics:<key:"_sampling_priority_v1" value:1 > metrics:<key:"php.compilation.total_time_ms" value:52.085 > metrics:<key:"process_id" value:12206 > type:"web"
```

### PHP version

```
PHP 8.0.17 (cli) (built: Jan  4 2023 15:54:10) ( NTS DEBUG )
Copyright (c) The PHP Group
Zend Engine v4.0.17, Copyright (c) Zend Technologies
    with ddtrace v1.0.0-nightly, Copyright Datadog, by Datadog
```

### Tracer version

dd-trace-php `afe28a4da3b3ad66b2e00bb6b48dd4d5571e4077`

### Installed extensions

<details>
<summary>[PHP Modules]</summary>

```
[PHP Modules]
apcu
ast
Core
ctype
curl
date
ddtrace
dom
FFI
fileinfo
filter
ftp
hash
iconv
json
libxml
mbstring
mcrypt
memcached
mongodb
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_pgsql
pdo_sqlite
Phar
posix
readline
redis
Reflection
session
SimpleXML
sockets
SPL
sqlite3
standard
tokenizer
xml
xmlreader
xmlwriter
zip
zlib

[Zend Modules]
ddtrace
```
</details>

### OS info

From the Docker:

```
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
```

#### Output of phpinfo() (ddtrace >= 0.47.0)

Straight from the Docker image mentioned in `CONTRIBUTING.md`

Let me know if you need more information.

Internal Support Ticket: 1066454

### PR DESCRIPTION
**Pull request by bwoebi**
### Description

Fixes #1859.

### Readiness checklist
- [ ] (only for Members) Changelog has been added to the release document.
- [ ] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: zend_abstract_interface/interceptor/php8/resolver.c [hunk 1]
```diff
@@ -5,8 +5,28 @@ static void zai_interceptor_function_declared(zend_op_array *op_array, zend_stri
     zai_hook_resolve_function((zend_function *)op_array, name);
 }
 
+#if PHP_VERSION_ID < 80300
```
Comment (morrisonlevi) [inline zend_abstract_interface/interceptor/php8/resolver.c:8]: If this is to work around a bug on two minor releases, why isn't this restricted to just those versions? This patch will also apply to 8.2.2, yes?

**Review by morrisonlevi — COMMENTED** [review]
Is it feasible to add a test for this? I don't understand the bug yet so I don't know if this fixes it.

File: zend_abstract_interface/interceptor/php8/resolver.c (hunk 1 above)
Comment (bwoebi) [reply zend_abstract_interface/interceptor/php8/resolver.c:8]: It will. I wasn't sure whether a version check here is worth it, but probably I should just do it.

**Comment by bwoebi** [comment]
Not sure exactly how to add a test for it, it bugs when the allocation by chance returns a non-zero value. Seems pretty non-deterministic to me.

File: zend_abstract_interface/interceptor/php8/resolver.c (hunk 1 above)
Comment (bwoebi) [reply zend_abstract_interface/interceptor/php8/resolver.c:8]: Done :-)

**Review by morrisonlevi — APPROVED** [review]
